### ISSUE TITLE
Transpiling CCXT to C++ for C++ support

### ISSUE THREAD

**Issue by MrMechanical** [ccxt/ccxt#10661]
I am attempting to transpile CCXT to C++ for maximum performance. I am wondering what is the best way to achieve this? My current attempt (in progress) is using Cython to convert the Python version of CCXT to C++. Perhaps my approach is not the best, I am curious what is the most optimal way of achieving this. I am aiming for maximum performance. Maybe transpilers through javascript or php could provide a superior result, though I wasn't able to find any transpilers for javascript->C++ or php->C++ that would be suitable. Thank you for any feedback or help.

Kind regards,
MrMechanical

**Comment by kroitor** [comment]
Hi!

> My current attempt (in progress) is using Cython to convert the Python version of CCXT to C++. Perhaps my approach is not the best, I am curious what is the most optimal way of achieving this.

Cython will transpile the code in a shallow way, that means a lot of redundancy in the resulting C/C++ code. Using Cython is one of the approaches, but probably not the best one if you're aiming for max performance.

The max-performance approach is to rewrite the base code in C++ manually and also rewrite the CCXT's built-in transpiler to support C++.

See these other attempts on adding the support for compiled languages:

- https://github.com/ccxt/ccxt/issues/1952
- https://github.com/prompt-cash/ccxt-go
- https://github.com/ccxt/ccxt/issues/2709
- https://github.com/irbis-labs/ccxt-rs

**Comment by ttodua** [comment]
@kroitor  can you mention why it's worth c++ version and  are those performance advantages any of worth over i.e. nodejs version?

**Comment by kroitor** [comment]
@ttodua yeah, you won't have the overhead of node.js, aggressive reference counting, garbage collection, and all together that should speed things up for you pretty significantly, or, put otherwise, it will make your software less demanding for resources, and therefore you could run more instances of it on the same machine, meaning better scalability, less latency, etc.

**Comment by ld2018** [comment]
Did anyone manage to write this in C++ yet?

**Comment by kroitor** [comment]
@ld2018 not yet, but we're making progress little by little, hope to move to a type-strict base language this year, and from there the path to supporting compiled languages should become easier.

**Comment by Serpens66** [comment]
Max performance is especially important for websocket. I don't know what the current ccxt-pro websocket code looks like, but I already have my own quite efficient python code for websocket and it is by far not fast enough to eg. handle a subscription from all ~1600 orderbooks from binance.
So for websocket one needs to utilize all CPUs from a machine (which python is bad at) and need fast code execution to sort the orderbook update to the own copy. 

Since many libraries for python are written in C++, I think it would be great if the Python version from ccxt would have C++ code for the websocket stuff.

**Comment by FreakieHuang** [comment]
> Hi!
> 
> > My current attempt (in progress) is using Cython to convert the Python version of CCXT to C++. Perhaps my approach is not the best, I am curious what is the most optimal way of achieving this.
> 
> Cython will transpile the code in a shallow way, that means a lot of redundancy in the resulting C/C++ code. Using Cython is one of the approaches, but probably not the best one if you're aiming for max performance.
> 
> The max-performance approach is to rewrite the base code in native C++ manually (easier said than done) and also rewrite the CCXT's built-in transpiler to support C++. This is on our roadmap for 2022.
> 
> See these other attempts on adding the support for compiled languages:
> 
> * [[New Language] Go #1952](https://github.com/ccxt/ccxt/issues/1952)
> * https://github.com/prompt-cash/ccxt-go
> * [[New Language] Rust #2709](https://github.com/ccxt/ccxt/issues/2709)
> * https://github.com/irbis-labs/ccxt-rs

Hi, how is C++ support going on?

**Comment by akurniawan** [comment]
Hi @kroitor, following up here regarding the move to a type-script base language such as C++, is there any updates on that?

**Comment by vitomarinovic** [comment]
Any updates on this?
